### ISSUE TITLE
Over zealous masquerading

### ISSUE THREAD

**Issue by stuarta** [firewalld/firewalld#501]
When masquerading is enabled on a zone containing the external interface for a system, a masquerade rule is created that enables masquerading for all traffic that is not on the loopback.

```
Chain POST_FedoraWorkstation_allow (1 references)
    pkts      bytes target     prot opt in     out     source               destination         
     953    62484 MASQUERADE  all  --  *      !lo     0.0.0.0/0            0.0.0.0/0
```

Let me explain the setup, and why this behaviour is wrong.

I have libvirt setup with a routed network (so libvirt does no masquerade / NAT for that network) and VM's on that network which should communicate directly with each other.

When in operation this creates the following network layout
```
enp0s31f6 -> virbr0 -> vnet0 (VM1)
enp0s31f6 -> virbr0 -> vnet1 (VM2)
```
`virbr0` has the ip address `192.168.44.1`, VM1 has `192.168.44.21`, VM2 has `192.168.44.22`.
`enp0s31f6` is in the FedoraWorkstation zone, on which I have enabled masquerade in order to masquerade any traffic *exiting* `enp0s31f6`, so that traffic returns correctly.

The problem begins with traffic between VM1 and VM2. I've run tcpdump and then ping between the VM's and instead of the traffic coming from the VM address, it comes from the `virbr0` address.

This then breaks things like openshift, which require and validate that the connections between the nodes are coming from where they are supposed to.

I've had to work around this by dropping the masquerade on the zone, and using direct rules to add back the NAT for any traffic *exiting* `enp0s31f6`. This is how I expect it should actually work

**Comment by erig0** [comment]
You can use a rich rule to do the masquerading. Something like the following should work.
```
# firewall-cmd --zone=FedoraWorkstation --add-rich-rule='rule family=ipv4 source address=192.168.44.0/24 destination not address=192.168.44.0/24 masquerade'
```

That'll result in the following iptables rule:
```
# iptables -v -t nat -L |grep -i masq
    0     0 MASQUERADE  all  --  any    !lo     192.168.44.0/24     !192.168.44.0/24     
```

**Comment by stuarta** [comment]
@erig0  That's what I've had to do, as mentioned in the workaround in the original problem description. Change the zone back to not masquerade, and add explicit masquerade rules for the subnet in question.

**Comment by erig0** [comment]
> @erig0 That's what I've had to do, as mentioned in the workaround in the original problem description. Change the zone back to not masquerade, and add explicit masquerade rules for the subnet in question.

I was pointing out that it can be done with a rich rule.

What do the routes look like for the container? Odd that the traffic between containers is "exiting" enp0s31f6. The masquerade rule is scoped by the interface, see POSTROUTING_ZONES_IFACES.

**Comment by stuarta** [comment]
@erig0 There is no chain POSTROUTING_ZONES_IFACES, the only rules setup by firewalld here for masquerading is the one from the original problem description which only has the restriction `!lo`

Therefore the rule matches as it transits out of the virbr up to the vnet device of the VM, and that's where it's getting the IP from.

What routes from the VM are you after? IIRC they all only have a default route out their `eth0`

**Comment by erig0** [comment]
> @erig0 There is no chain POSTROUTING_ZONES_IFACES, the only rules setup by firewalld here for masquerading is the one from the original problem description which only has the restriction `!lo`

My bad. That chain only exists in newer versions. If you change the default zone to something else (e.g. drop), then explicitly assign your interfaces to FedoraWorkstation - does it work?

**Comment by stuarta** [comment]
@erig0 Done some further testing based on your suggestions here.

- With the default zone set to block, the external nic added back to FedoraWorkstation zone, and masquerading enabled on FedoraWorkstation zone, the traffic between the 2 VM's is correctly left alone.
- With the default zone set to FedoraWorkstation, the external nic in FedoraWorkstation zone, and masquerading enabled on FedoraWorkstation zone, the traffic between the 2 VM's is incorrectly NAT'd to the ip address on the `virbr` which the 2 VM's are connected to.

If you want them, i dumped the iptables rules and nat table rules in both setups. Regardless, it's easy to reproduce locally. All I needed was an f30 host, and 2x f30 VM's on a routed libvirt network

**Comment by erig0** [comment]
> @erig0 Done some further testing based on your suggestions here.
> 
>     * With the default zone set to block, the external nic added back to FedoraWorkstation zone, and masquerading enabled on FedoraWorkstation zone, the traffic between the 2 VM's is correctly left alone.
> 
>     * With the default zone set to FedoraWorkstation, the external nic in FedoraWorkstation zone, and masquerading enabled on FedoraWorkstation zone, the traffic between the 2 VM's is incorrectly NAT'd to the ip address on the `virbr` which the 2 VM's are connected to.

Based on this I think we can conclude this is another symptom of #258 and #441. Without upgrading firewalld you have multiple workarounds:
1. Change the default zone (as you've done above)
2. Change the `--set-target` of the zone virbr0 is assigned to - something other than `default`. Or simply use the trusted zone.

This was fixed in 0.7.0 and the upcoming 0.6.5.

**Comment by stuarta** [comment]
FYI. I had tried putting the `virbr` into the trusted zone and it made no difference.
